Create 'write effective documentation' pattern (#147)
HO-CTO/engineering-guidance-and-standards: docs/patterns/write-clear-documentation.md

---
layout: sub-navigation
order: 1

> title: Write clear documentation

Reviewer: Clear is a subjective term and poorly defined, possibly consider understandable or effective?

Author: Good shout, 'effective' is better

The benefits of well-written documentation are often understated, but can transform the experience for engineers and users. If it is easy for readers to understand how to use a product, it improves self-service, so they will not need to contact support for help as often. They can become experts at using a product on their own, reducing the amount of time and money spent on troubleshooting, and improving productivity.

Following this guidance, and the [GOV.UK guidance on how to write well for your audience, including specialists.](https://www.gov.uk/guidance/content-design/writing-for-gov-uk), will help you to write clear and accessible technical documentation for your audience.

---

## Solution

- [Start with User Needs](#start-with-user-needs)
- [Make your content accessible and inclusive](#make-your-content-accessible-and-inclusive)
- [Write clear content](#write-clear-content)

### Start with User Needs

"_If you don't know what the user needs are, you won't build (or write) the right thing_"

> Engineers are often users too! The key to successful documentation is that it meets the needs of the audience in a way that they understand. Taking a user-centred approach to how you design and write your documentation is a great way to make it successful.

Reviewer: ```suggestion
The key to successful documentation is that it meets the needs of the audience in a way that they understand. Taking a user-centred approach to how you design and write your documentation is a great way to make it successful. Engineers are often users too, make sure you have identified all the users of your documentation when assessing their needs.
```

#### How to do this

- Write down who your users are and consider how technical they are
- Think about what they need to do with the thing you are documenting
- Break your content into conceptual, task based, and reference information

> - Use [accessibility personas](https://alphagov.github.io/accessibility-personas/) to help you consider users with access needs

Reviewer: ```suggestion
- Use [accessibility personas](https://alphagov.github.io/accessibility-personas/) to help you consider users with access needs
- Use an iterative approach to get feedback so that you verify what you write meets the identified needs.
```

Author: Slightly rewritten but have added bullet on this point

#### Examples of user needs

> |As a Site Reliability Engineer in a product team <br><br> I need to understand the steps to set up Monitoring as Code (MaC) <br><br> So that I do not need to contact the platform team|As a System Administrator <br><br> I need to set up access levels for system users <br><br> So that the right people have the correct rights| As a Support Analyst <br><br> I need to understand what I need to do if something goes wrong <br><br> So that I can get the system back up and running quickly|

Reviewer 2: I think the user story examples are harder to read when in the table; for myself, it was not immediately clear how to read it.

A more clear approach (for both the output and the raw markdown) may be to use lists, such as below. 

```suggestion
- User need example 1 
  - As a Site Reliability Engineer in a product team
  - I need to understand the steps to set up Monitoring as Code (MaC)
  - So that I do not need to contact the platform team
- User need example 2
  - As a System Administrator
  - I need to set up access levels for system users
  - So that the right people have the correct rights
- User need example 3
  - As a Support Analyst
  - I need to understand what I need to do if something goes wrong
  - So that I can get the system back up and running quickly
```

#### Make images and diagrams accessible

Do not rely on the visual element of pictures to convey meaning, that meaning should be described with text. If you do use images or diagrams, make them simple and clear, and include alt-text so that screen readers can describe them for your readers.

The Home Office Accessibility team have published guidance on [creating accessible images](https://design.homeoffice.gov.uk/accessibility/images).

#### Use inclusive language

The Home Office accessibility team have published guidance on [how to write in a way that is compassionate and respectful](https://design.homeoffice.gov.uk/accessibility/inclusive-language). You should follow this guidance, and also remember that there are common technical concepts that should be deprecated, below are some examples (but the list is not exhaustive).

| Example | Use instead |
|:---|:---|
| Whitelist | Allowlist |
| Blacklist | Denylist or blocklist |
| Master | Primary or main |
| Slave | Secondary |
| Man hours | Person hours or engineer hours |
| Sanity check | Quick check or confidence check |
| Dummy value | Placeholder value |

> ### Write clear content

Reviewer: ```suggestion
### Reduce readers' cognitive load
```

Author: I have gone with 'Follow content writing best practices'

> Complexity is confusing, overwhelming and can waste time. [People with specialist knowledge prefer clear content](https://gds.blog.gov.uk/2014/02/17/guest-post-clarity-is-king-the-evidence-that-reveals-the-desperate-need-to-re-think-the-way-we-write/). Using plain English and structuring your documentation will result in clearer content.

Reviewer: ```suggestion
Complexity is confusing, overwhelming and can waste time. [People with specialist knowledge prefer clear content](https://gds.blog.gov.uk/2014/02/17/guest-post-clarity-is-king-the-evidence-that-reveals-the-desperate-need-to-re-think-the-way-we-write/). Using plain English and carefully structuring your documentation will reduce the cogitive load required to understand your documentation.

Not all complexity can be avoided. If you're producing documentation then there is likely some necessesary complexity arising from the subject of the document. The goal is to ensure that the writing style and presentation of the document aid the reader in understanding the subject.
```

#### Write in plain English

- Use the active voice and the present tense ([NHS Digital have useful guidance on the active voice](https://digital.nhs.uk/about-nhs-digital/corporate-information-and-documents/our-style-guidelines/content-style-guide/a-to-z-of-writing#active,-passive))
- Look for a simpler word and try to convey meaning with the fewest words possible
- Break up long sentences and paragraphs
- Remove any negative contractions, like 'isn't' or 'don't'
- The [Home Office content style guide](https://design.homeoffice.gov.uk/content-style-guide) extends the [GOV.UK style guide](https://www.gov.uk/guidance/style-guide) for some specific Home Office cases. Use these guides to ensure you use consistent language

#### How to structure your documentation

- Check that your title describes the content beneath it
- Use clear headings and subheadings

> - Use bulleted and numbered lists

Reviewer: ```suggestion
- Use bulleted and numbered lists where appropriate to group concepts and break up the content.
```

- Say in your content who it is for

> #### Iterate and improve your content alongside your product or service

Reviewer: This seems like it should be its own section, not grouped into the clear documentation section

Author: This makes more sense now that the section is labelled as 'follow content writing best practices'

---
layout: sub-navigation
order: 1
title: Write effective documentation
date: git Last Modified
tags:
- Ways of working
- Documentation
- Accessibility
---

{% import "_includes/macros/tags.njk" as pageTag %}

Last updated: {{ page.date | postDate }}
{{ pageTag.tags(tags)  }}

Technical documentation explains how a product works, and its intended readers can range anywhere from software developers to non-technical users. It can take a number of forms, such as an onboarding guide or API documentation. Typically, technical writers and engineers create this type of documentation.

> The benefits of well-written documentation are often understated, but can transform the experience for engineers and users. If it is straightforward for readers to understand how to use a product, it improves self-service, so they will not need to contact support for help as often. They can become experts at using a product on their own, reducing the amount of time and money spent on troubleshooting, and improving productivity.

Reviewer 3: A suggestion if you would like to be more assertive on this point:

```suggestion
Well-written documentation can transform the experience for engineers and users.
```

Following this guidance, and the [GOV.UK guidance on how to write well for your audience, including specialists.](https://www.gov.uk/guidance/content-design/writing-for-gov-uk), will help you to write effective and accessible technical documentation for your audience.

---

## Solution

- [Start with User Needs](#start-with-user-needs)
- [Make your content accessible and inclusive](#make-your-content-accessible-and-inclusive)
- [Follow content writing best practices](#follow-content-writing-best-practices)

### Start with User Needs

> "_If you don't know what the user needs are, you won't build (or write) the right thing_"

Reviewer 3: Does this quote need a citation?

Author: Ha yes, it probably should as a slightly cheeky rewrite of the first of the Government Design Principles here https://www.gov.uk/guidance/government-design-principles

> The key to successful documentation is that it meets the needs of the audience in a way that they understand. Taking a user-centred approach to how you design and write your documentation is a great way to make it successful. Engineers are often users too, identify all the potential readers of your documentation when assessing their needs.

Reviewer 3: ```suggestion
The key to successful documentation is that it meets the needs of the audience in a way that they understand. Taking a user-centred approach to how you design and write your documentation is a great way to make it effective. Engineers are often users too, identify all of the potential readers of your documentation when assessing user needs.
```

- Avoid 'Frequently Asked Questions' sections (FAQs)
- Spell out acronyms in the first instance

#### Make links accessible

See the Home Office accessibility team's guidance to [learn how to write visible and descriptive links that behave in an unsurprising way](https://design.homeoffice.gov.uk/accessibility/links).

Use links to supporting documentation to avoid duplicating content, especially when referring to third-party services. It is easier to maintain if they change their processes. For example: 'See the GitHub documentation for more information on [generating an SSH key](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent).'

#### Make images and diagrams accessible

Do not rely on the visual element of pictures to convey meaning, that meaning should be described with text. If you do use images or diagrams, make them simple and clear, and include alt-text so that screen readers can describe them for your readers.

The Home Office Accessibility team have published guidance on [creating accessible images](https://design.homeoffice.gov.uk/accessibility/images).

#### Use inclusive language

> The Home Office accessibility team have published guidance on [how to write in a way that is compassionate and respectful](https://design.homeoffice.gov.uk/accessibility/inclusive-language). You should follow this guidance, and also remember that there are common technical concepts that should be deprecated, below are some examples (but the list is not exhaustive).

Reviewer 3: ```suggestion
The Home Office accessibility team have published guidance on [how to write in a way that is compassionate and respectful](https://design.homeoffice.gov.uk/accessibility/inclusive-language). You should follow this guidance, and also remember that there are antiquated instances of technical terminology that should be deprecated, below are some examples (but the list is not exhaustive).
```
